zerver/actions/message_send.py

```diff
@@ -1525,19 +1527,60 @@ def validate_stream_id_with_pm_notification(
     return stream
 
 
-def check_private_message_policy(
-    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile]
+def check_can_send_direct_message(
+    realm: Realm, sender: UserProfile, recipient_users: Sequence[UserProfile], recipient: Recipient
 ) -> None:
-    if realm.private_message_policy == PrivateMessagePolicyEnum.DISABLED:
-        if sender.is_bot or (
-            len(user_profiles) == 1 and (user_profiles[0].is_bot or user_profiles[0] == sender)
-        ):
-            # We allow direct messages only between users and bots or to oneself,
-            # to avoid breaking the tutorial as well as automated
-            # notifications from system bots to users.
-            return
+    if sender.is_bot:
+        return
+
+    if all(user_profile.is_bot or user_profile.id == sender.id for user_profile in recipient_users):
+        return
+
+    direct_message_permission_group = realm.direct_message_permission_group
+
+    # TODO: This check is inefficient; we should in the future be able to cache
+    # on the Huddle object whether the conversation already exists, likely in the
+    # form of a `first_message_id` field, and be able to save doing this check in the
+    # common case that this is not the first message in a conversation.
```

Comment:
This comment is misplaced, since it's about the initiator group. Fixed this and also some general confusion where you're not using the `recipient` parameter passed  into the function in obvious places.

``` diff
diff --git a/zerver/actions/message_send.py b/zerver/actions/message_send.py
index a38584d495..465d5d926c 100644
--- a/zerver/actions/message_send.py
+++ b/zerver/actions/message_send.py
@@ -1538,10 +1538,6 @@ def check_can_send_direct_message(
 
     direct_message_permission_group = realm.direct_message_permission_group
 
-    # TODO: This check is inefficient; we should in the future be able to cache
-    # on the Huddle object whether the conversation already exists, likely in the
-    # form of a `first_message_id` field, and be able to save doing this check in the
-    # common case that this is not the first message in a conversation.
     if (
         not hasattr(direct_message_permission_group, "named_user_group")
         or direct_message_permission_group.named_user_group.name != SystemGroups.EVERYONE
@@ -1560,6 +1556,11 @@ def check_can_send_direct_message(
     ):
         if is_user_in_group(direct_message_initiator_group, sender):
             return
+
+        # TODO: This check is inefficient; we should in the future be able to cache
+        # on the Huddle object whether the conversation already exists, likely in the
+        # form of a `first_message_id` field, and be able to save doing this check in the
+        # common case that this is not the first message in a conversation.
         if recipient.type == Recipient.PERSONAL:
             recipient_user_profile = recipient_users[0]
             previous_messages_exist = (
@@ -1568,15 +1569,15 @@ def check_can_send_direct_message(
                     recipient__type=Recipient.PERSONAL,
                 )
                 .filter(
-                    Q(sender=sender, recipient__type_id=recipient_user_profile.id)
-                    | Q(sender=recipient_user_profile, recipient=sender.recipient)
+                    Q(sender=sender, recipient=recipient)
+                    | Q(sender=recipient_user_profile, recipient_id=sender.recipient_id)
                 )
                 .exists()
             )
         else:
+            assert recipient.type == Recipient.DIRECT_MESSAGE_GROUP
             previous_messages_exist = Message.objects.filter(
                 realm=realm,
-                recipient__type=Recipient.DIRECT_MESSAGE_GROUP,
                 recipient=recipient,
             ).exists()
         if not previous_messages_exist:
```